plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -1279,6 +1320,121 @@ public static Optional<ReaderColumns> projectColumns(List<IcebergColumnHandle> c
     }
 
+    /**
+     * Creates a set of sufficient columns for the input projected columns and prepares a mapping between the two.
+     * For example, if input {@param columns} include columns "a.b" and "a.b.c", then they will be projected
+     * from a single column "a.b".
+     */
+    private static Optional<ReaderColumns> projectSufficientColumns(List<IcebergColumnHandle> columns)
+    {
+        requireNonNull(columns, "columns is null");
+
+        if (columns.stream().allMatch(IcebergColumnHandle::isBaseColumn)) {
+            return Optional.empty();
+        }
+
+        ImmutableBiMap.Builder<DereferenceChain, IcebergColumnHandle> dereferenceChainsBuilder = ImmutableBiMap.builder();
+
+        for (IcebergColumnHandle column : columns) {
+            DereferenceChain dereferenceChain = new DereferenceChain(column.getBaseColumnIdentity(), column.getPath());
+            dereferenceChainsBuilder.put(dereferenceChain, column);
+        }
+
+        BiMap<DereferenceChain, IcebergColumnHandle> dereferenceChains = dereferenceChainsBuilder.build();
+
+        List<ColumnHandle> sufficientColumns = new ArrayList<>();
+        ImmutableList.Builder<Integer> outputColumnMapping = ImmutableList.builder();
+
+        Map<DereferenceChain, Integer> pickedColumns = new HashMap<>();
+
+        // Pick a covering column for every column
+        for (IcebergColumnHandle columnHandle : columns) {
+            DereferenceChain dereferenceChain = dereferenceChains.inverse().get(columnHandle);
+            List<DereferenceChain> orderedPrefixes = dereferenceChain.getOrderedPrefixes();
+            DereferenceChain chosenColumn = null;
+
+            // Shortest existing prefix is chosen as the input.
+            for (DereferenceChain prefix : orderedPrefixes) {
+                if (dereferenceChains.containsKey(prefix)) {
+                    chosenColumn = prefix;
+                    break;
+                }
+            }
+
+            checkState(chosenColumn != null, "chosenColumn is null");
+            int inputBlockIndex;
+
+            if (pickedColumns.containsKey(chosenColumn)) {
+                // Use already picked column
+                inputBlockIndex = pickedColumns.get(chosenColumn);
+            }
+            else {
+                // Add a new column for the reader
+                sufficientColumns.add(dereferenceChains.get(chosenColumn));
+                pickedColumns.put(chosenColumn, sufficientColumns.size() - 1);
+                inputBlockIndex = sufficientColumns.size() - 1;
+            }
+
+            outputColumnMapping.add(inputBlockIndex);
+        }
+
+        return Optional.of(new ReaderColumns(sufficientColumns, outputColumnMapping.build()));
+    }
+
+    private static Optional<org.apache.parquet.schema.Type> getColumnType(IcebergColumnHandle column, Map<Integer, org.apache.parquet.schema.Type> parquetIdToField)
```

Comment:
Don't you already have all the types in your `parquetIdToField` map already?

Not sure why you need this method instead of just doing `Optional.ofNullable(parquetIdToField.get(column.getId())`

Reply:
for a dereference field `a.c`,

`parquetIdToField.get(column.getId())` returns
```
message table {
    optional int32 c = 4;
  }
```

We would require a type with a fully qualified path for the dereference field like below

```
  message table {
    optional group a {
      optional int32 c = 4;
    }
  }
```